rllib/utils/exploration/parameter_noise.py

```diff
@@ -0,0 +1,386 @@
+        # Is this the first forward pass in the new episode? If yes, do the
+        # noise re-sampling and add to weights.
+        if self.episode_started:
+            self._delayed_on_episode_start(tf_sess)
+
+        explore = explore if explore is not None else \
+            self.policy_config["explore"]
+        # Add noise if necessary.
+        if explore and not self.weights_are_currently_noisy:
+            self._add_stored_noise(tf_sess=tf_sess)
+        # Remove noise if necessary.
+        elif not explore and self.weights_are_currently_noisy:
+            self._remove_noise(tf_sess=tf_sess)
+
+    @override(Exploration)
+    def get_exploration_action(self,
+                               *,
+                               distribution_inputs,
+                               action_dist_class,
+                               timestep,
+                               explore=True):
+        # Use our sub-exploration object to handle the final exploration
+        # action (depends on the algo-type/action-space/etc..).
+        return self.sub_exploration.get_exploration_action(
+            distribution_inputs=distribution_inputs,
+            action_dist_class=action_dist_class,
+            timestep=timestep,
+            explore=explore)
+
+    @override(Exploration)
+    def on_episode_start(self,
+                         policy,
+                         *,
+                         environment=None,
+                         episode=None,
+                         tf_sess=None):
+        # We have to delay the noise-adding step by one forward call.
+        # This is due to the fact that the optimizer does it's step right
+        # after the episode was reset (and hence the noise was already added!).
+        # We don't want to update into a noisy net.
+        self.episode_started = True
+
+    def _delayed_on_episode_start(self, tf_sess):
+        # Sample fresh noise and add to weights.
+        if self.default_explore:
+            self._sample_new_noise_and_add(tf_sess=tf_sess, override=True)
+        # Only sample, don't apply anything to the weights.
+        else:
+            self._sample_new_noise(tf_sess=tf_sess)
+        self.episode_started = False
+
+    @override(Exploration)
+    def on_episode_end(self,
+                       policy,
+                       *,
+                       environment=None,
+                       episode=None,
+                       tf_sess=None):
+        # Remove stored noise from weights (only if currently noisy).
+        if self.weights_are_currently_noisy:
+            self._remove_noise(tf_sess=tf_sess)
+
+    @override(Exploration)
+    def postprocess_trajectory(self, policy, sample_batch, tf_sess=None):
+        noisy_action_dist = noise_free_action_dist = None
+        # Adjust the stddev depending on the action (pi)-distance.
+        # Also see [1] for details.
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
```

Comment:
```suggestion
        noisy_action_dist = policy.compute_action_distribution(
```

Reply:
Hmm, this is not good. The Policy.compute_action_distribution method was not even part of this PR. Let me see, whether using this method can be avoided for now entirely.

Reply:
It's only used in ParameterNoise::post process_traj. I'll rename and re-purpose it as you suggested, but it'll not be used (actually called anywhere) till the next PR.